### ISSUE TITLE
TypeScript support + support for 0-sized rectangles

### ISSUE THREAD

**Issue by jurca** [lovasoa/rectangle-overlap#1]
Hello,

would you be willing to support TypeScript ecosystem? There are various options, however, I think it would be best for future maintainability to port the code to TypeScript instead of adding separate type declaration files.

I am willing to do all the work myself and send a pull request.

Would that be OK with you, or would you rather go for a different option?

I would also like to add a distinction between non-intersecting rectangles and a 0-sized rectangle inside another any-sized rectangle by returning `null` instead of `0` when rectangles are not intersecting. Would that be OK with you?

**Comment by lovasoa** [comment]
Hello, and thank you for your interest in this library ! 

Converting the code to typescript is ok, and I would merge such a pull request. Don't forget to make the necessary changes to `package.json`.

The second change is ok too, but you should document it both in doc comments and in the README.

**Comment by jurca** [comment]
A little hiccup - would you be OK with changing the licence to L-GPL, allowing the library to be used in commercial projects (without having to license them as GPL as well)?

**Comment by lovasoa** [comment]
Yes, It's OK too.

**Comment by jurca** [comment]
I have progressed with the package update (you can see the current state here: https://github.com/jurca/rectangle-overlap), however, while adding unit tests, I have found a bug:

```javascript
var area = require("rectangle-overlap")
var r1 = {x: 0, y: 0, w: 100, h: 100}
var r2 = {x: 10, y: -100, w: 1, h: 101}
area(r1.x, r1.y, r1.w, r1.h, r2.x, r2.y, r2.w, r2.h)
```

The result is `90`, however, the correct result would be `1`.

This got me thinking... we are doing breaking changes anyway. Maybe, instead of just fixing this, we could make it a little more useful - instead of just returning the overlapping area, it would return the intersection as a rectangle:

```javascript
> var calculateOverlap = require("rectangle-overlap")
> var r1 = {x: 0, y: 0, w: 100, h: 100}
> var r2 = {x: 10, y: -100, w: 1, h: 101}
> calculateOverlap(r1.x, r1.y, r1.w, r1.h, r2.x, r2.y, r2.w, r2.h)
{
  x: 10,
  y: 0,
  w: 1,
  h: 1,
}
```

This would also make the code easier to test&debug. What do you think?

**Comment by lovasoa** [comment]
We could even have a rectangle class with an area method, take two `Rectangle`s in and output one `Rectangle`.

**Comment by jurca** [comment]
Sounds great, however, I would prefer not to force the user of the library to construct the rectangle instances. How about we expect rectangle-shaped objects as input and output a `Rectangle` class instance instead? That would keep the library convenient to use.

I propose the following shape of the `Rectangle` class:

```typescript
class Rectangle {
  readonly x: number
  readonly y: number
  readonly w: number
  readonly width: number
  readonly h: number
  readonly height: number
  readonly area: number
}
```

and the following usage:

```typescript
> var r1 = {x: 0, y: 0, w: 100, h: 100}
> var r2 = {x: 10, y: -100, w: 1, h: 101}
> calculateOverlap(r1, r2)
Rectangle {
  x: 10
  y: 0
  w: 1
  width: 1
  h: 1
  height: 1
  area: 1
}
```

**Comment by lovasoa** [comment]
I'd prefer having only `width` and `height`, without `w`and `h`... What do you think ?

**Comment by jurca** [comment]
I'm OK with that, I included them only because I suspected you might prefer to have them.

I'll fix and update the implementation.

**Comment by jurca** [comment]
Here's the pull request: https://github.com/lovasoa/rectangle-overlap/pull/2

Could you, please, review it? Thanks!

**Comment by lovasoa** [comment]
The new version was published to npm as 2.0.0: https://www.npmjs.com/package/rectangle-overlap/v/2.0.0
